ledger/participant-integration-api/src/test/suite/scala/platform/store/backend/UpdateToDbDtoSpec.scala

```diff
@@ -0,0 +1,1191 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.backend
+
+import java.time.{Duration, Instant}
+
+import com.daml.daml_lf_dev.DamlLf
+import com.daml.ledger.api.domain
+import com.daml.ledger.api.v1.event.{CreatedEvent, ExercisedEvent}
+import com.daml.ledger.participant.state.v1.{
+  ApplicationId,
+  CommandId,
+  Configuration,
+  DivulgedContract,
+  Offset,
+  ParticipantId,
+  Party,
+  RejectionReasonV0,
+  SubmissionId,
+  SubmitterInfo,
+  TimeModel,
+  TransactionId,
+  TransactionMeta,
+  Update,
+  WorkflowId,
+}
+import com.daml.lf.crypto
+import com.daml.lf.data.{Ref, Time}
+import com.daml.lf.ledger.EventId
+import com.daml.lf.transaction.BlindingInfo
+import com.daml.lf.transaction.test.TransactionBuilder
+import com.daml.lf.value.Value
+import com.daml.logging.LoggingContext
+import com.daml.platform.store.appendonlydao.JdbcLedgerDao
+import com.daml.platform.store.appendonlydao.events.Raw.TreeEvent
+import com.daml.platform.store.appendonlydao.events.{
+  CompressionStrategy,
+  ContractId,
+  Create,
+  Exercise,
+  FieldCompressionStrategy,
+  LfValueSerialization,
+  Raw,
+}
+import com.daml.platform.store.dao.DeduplicationKeyMaker
+import com.google.protobuf.ByteString
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AnyWordSpec
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class UpdateToDbDtoSpec extends AnyWordSpec with Matchers {
+  // DbDto case classes are not comparable, because the contains Arrays.
+  // The DbDto trait mixes in SomeArrayEquals, but this does not help as some DTOs contain Arrays within Options.
```

Comment:
rant
actually this also means wartremover is a bit shortsighted
also seem to mean like what is ok for the scala-lib, is not ok for us to do :( 
also means like SomeArrayEquals has some limitations
all this make me wish just using plain case classes with whatever we want again :man_facepalming: 
/rant

Reply:
>  actually this also means wartremover is a bit shortsighted

I can't blame wartremover here at all; we just aren't doing what really must be done to solve this, which is to avoid `Object#equals` and its proxy `==` altogether. (Note that we do not enable [the `Equals` wart](http://www.wartremover.org/doc/warts.html#equals).)

`==` is not type-safe and there is no way to make it type-safe by subtraction; if you want a safe equals, but not to define full typeclasses as Scalaz, Cats, and Spire do, the only way is [still by addition](https://github.com/japgolly/univeq#univeq).

(I come not to suppress your rant but to amplify it to its natural conclusion 😄 )

>  also means like SomeArrayEquals has some limitations

Absolutely. As a hack to make particular speedy case classes comparable without filling out a bunch of buggy boilerplate, it's not flexible beyond what was apparent in those classes.